src/signer-protocol/src/main/java/ee/ria/xroad/signer/protocol/SignerClient.java

```diff
@@ -65,95 +72,191 @@ private SignerClient() {
+         * long request timeouts when the signer is not (yet) available, and to detect restarts.
+         *
+         */
+        private static volatile CompletableFuture<ActorRef> requestProcessor = null;
+        private static final Duration WATCH_DELAY = Duration.ofSeconds(1);
+        private static final int REF_GET_TIMEOUT = 7;
 
-    private static void verifyInitialized() {
-        if (actorSystem == null) {
-            throw new IllegalStateException("SignerClient is not initialized");
+        static ActorRef signerRef() {
+            final CompletableFuture<ActorRef> processor = requestProcessor;
+            if (processor == null) {
+                throw new IllegalStateException("SignerClient is not initialized");
+            }
+            try {
+                return processor.get(REF_GET_TIMEOUT, TimeUnit.SECONDS);
+            } catch (ExecutionException | TimeoutException | CancellationException e) {
+                throw new CodedException(X_INTERNAL_ERROR, e, "Signer is unreachable");
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new CodedException(X_INTERNAL_ERROR, e, "Request to signer was interrupted");
+            }
         }
-    }
 
-    private static CodedException connectionTimeoutException(Exception e) {
-        return new CodedException(X_HTTP_ERROR, e,
-                "Connection to Signer (port %s) timed out",
-                SystemProperties.getSignerPort());
-    }
+        static synchronized void init(ActorSystem system, String signerIpAddress) {
+            if (requestProcessor == null) {
+                requestProcessor = new CompletableFuture<>();
+                system.actorOf(Props.create(SignerWatcher.class, signerIpAddress));
+            }
+        }
+
+        private static synchronized void resetRequestProcessor(CompletableFuture<ActorRef> processor) {
+            if (requestProcessor != null && !requestProcessor.isDone()) {
+                requestProcessor.cancel(true);
+            }
+            requestProcessor = processor;
+        }
+
+        private long correlationId = 0;
+        private ActorRef signerRef;
+        private ActorSelection signer;
+        private final String signerIpAddress;
+
+        interface Watch {
+        }
 
+        SignerWatcher(String signerIpAddress) {
+            this.signerIpAddress = signerIpAddress;
+        }
+
+        @Override
+        public void preStart() {
+            signer = context().actorSelection(getSignerPath() + "/user/" + REQUEST_PROCESSOR);
+            self().tell(Watch.class, self());
+        }
+
+        @Override
+        public void postStop() {
+            if (signerRef != null) {
+                context().unwatch(signerRef);
+            }
+            resetRequestProcessor(null);
+        }
+
+        @Override
+        public void onReceive(final Object message) {
```

Comment:
The code related to SignerClient, SignerWatcher and signer requestProcessor tracking is pretty complex. I read it for some time and was barely able to grasp what is happening, and would be a bit scared if I had to update it.

I think ideally some sort of diagram would be great aid to help understanding the implementation, so that people can later update the implementation with better confidence. Perhaps it could show things like the actual client, SignerClient, SignerWatcher, Signer, and the different messages that flow between these (Watch, Identify, ActorIdentity, Terminated).

If not a diagram, then maybe consider adding some textual comments to make it easier to understand the mechanism.

Also behavior of `correlationId` and `signerRef` might benefit from more explanation.